controllers/bpfman-agent/cl_application_program.go

```diff
@@ -441,115 +427,99 @@ func (r *ClBpfApplicationReconciler) getProgState(prog *bpfmaniov1alpha1.ClBpfAp
+// ready. bpfman saves state in the BpfApplicationState object that controls
+// what needs to be done, so it is critical for each reconcile attempt to have
+// the updated information. However, it takes time for objects to be created or
+// updated, and for the API server to be able to return the update. When
+// waitForBpfAppStateUpdate gets the updated object, it also updates
+// r.currentAppState so the object can be used for subsequent operations (like a
+// status update).
+func (r *ClBpfApplicationReconciler) waitForBpfAppStateSpecUpdate(ctx context.Context) error {
+	r.Logger.V(1).Info("Start waitForBpfAppStateSpecUpdate()")
+	i := 0
+	err := wait.PollUntilContextTimeout(ctx, updateRetryInterval, updateTimeout, true,
+		func(ctx context.Context) (bool, error) {
+			i++
+			bpfAppState, err := r.getBpfAppState(ctx)
+			if err != nil {
+				// If we get an error, we'll just log it and keep trying.
+				r.Logger.V(1).Info("Error getting BpfApplicationState", "Attempt", i, "error", err)
+				return false, nil
+			}
+			if bpfAppState != nil {
+				r.Logger.Info("Found created BpfApplicationState Spec", "Attempt", i)
+				r.currentAppState = bpfAppState
+				return true, nil
+			}
+			r.Logger.V(1).Info("Didn't find new BpfApplicationState", "Attempt", i)
+			return false, nil
+		})
+	if err != nil {
+		return fmt.Errorf("failed to get updated BpfApplicationState spec.  Attempts: %d", i)
 	}
-
-	r.Logger.Info("Didn't find new BpfApplicationState", "Attempts", maxRetries)
-	return false, fmt.Errorf("failed to get new BpfApplicationState after %d retries", maxRetries)
+	return nil
 }
 
 // See waitForBpfAppStateUpdate() for an explanation of why this function is needed.
 func (r *ClBpfApplicationReconciler) waitForBpfAppStateStatusUpdate(ctx context.Context) error {
-	const maxRetries = 100
-	const retryInterval = 100 * time.Millisecond
-
-	var bpfAppState *bpfmaniov1alpha1.ClusterBpfApplicationState
-	var err error
-	r.Logger.Info("waitForBpfAppStateStatusUpdate()", "UpdateCount", r.currentAppState.Spec.UpdateCount,
-		"currentGeneration", r.currentAppState.GetGeneration())
-
-	for i := 0; i < maxRetries; i++ {
-		bpfAppState, _, err = r.getBpfAppState(ctx, false)
-		if err != nil {
-			// If we get an error, we'll just log it and keep trying.
-			r.Logger.Info("Error getting BpfApplicationState", "Attempt", i, "error", err)
-		} else if bpfAppState != nil && len(bpfAppState.Status.Conditions) > 0 &&
-			bpfAppState.Status.Conditions[0].Type == r.currentAppState.Status.Conditions[0].Type {
-			r.Logger.Info("Found new bpfAppState Status", "Attempt", i, "UpdateCount", bpfAppState.Spec.UpdateCount,
-				"currentGeneration", bpfAppState.GetGeneration())
-			r.currentAppState = bpfAppState
-			return nil
-		}
-		time.Sleep(retryInterval)
+	r.Logger.V(1).Info("Start waitForBpfAppStateStatusUpdate()")
+	i := 0
+	err := wait.PollUntilContextTimeout(ctx, updateRetryInterval, updateTimeout, true,
+		func(ctx context.Context) (bool, error) {
+			i++
```

Comment:
I don't think we need to use `i` anymore in here

Reply:
You are correct that we don't need it from a functional perspective, however, I'm currently logging a message that says how many iterations it took to get the updated CRD so we can track how it works on different/larger systems.  And, I'm not aware of another way to know how many polls were required.

Reply:
normally we don't track iterations just log when timeout condition happens tracking iterations might jist generate noise in the logs

Reply:
I think we can eventually get rid of this, but we weren't sure what was the right amount of time to wait.  So, I thought it would be good to do some more testing on larger deployments, and see exactly  how long it's taking for the updates to become available.